### ISSUE TITLE
nvim_buf_get_option for non-current buffer triggers redraw

### ISSUE THREAD

**Issue by blueyed** [neovim/neovim#8800]
```vim
call setqflist([{'filename': 'unlisted'}])
let unlisted = bufnr('^unlisted')

" Ok.
call nvim_buf_get_option(bufnr('%'), 'filetype')

" Triggers redraw!
call nvim_buf_get_option(unlisted, 'filetype')
```

Initially reported for neovim-python-client: https://github.com/neovim/python-client/issues/331.

This is annoying when e.g. a Python plugin uses this to filter buffers, e.g. in jedi-vim:
```python
    jedi.settings.additional_dynamic_modules = [
        b.name for b in vim.buffers if (
            b.name is not None and
            b.name.endswith('.py') and
            # XXX: triggers redraw!
            b.options['buflisted'])]
```

Might be related to this code (path), but have not investigated further: https://github.com/neovim/neovim/blob/9abe0bdb54a05932727665f68044ad23ed34fc30/src/nvim/option.c#L4768-L4773

**Comment by blueyed** [comment]
`getbufvar(unlisted, '&filetype')` is not affected.

**Comment by justinmk** [comment]
Isn't this the same as https://github.com/neovim/neovim/issues/3416 ? 

@blueyed Just for clarity: what's the use case for avoiding a redraw?

We can't know which `K_EVENT` needs a redraw or not.

The only approach I can think of is to debounce or "vsync" redraws.  This would require storing recent message(s) and 'showcmd' text as part of the logical screen, so that a deferred redraw is not destructive. cc @bfredl

**Comment by bfredl** [comment]
> We can't know which K_EVENT needs a redraw or not.

Just any (deferred) API call _per se_ shouldn't cause a redraw, I don't think `nvim_eval('1+1')` here would do so (correct me if I'm wrong), and if it doesn't neither should getting an option.

 I certainly agree messages management should be decoupled more from "vsyncs" (messages should just be some state drawn on a grid, just like a buffer window, which we can manipulate at will), but more short term, there probably actually is some state management in the get-buf-option code that unnecessarily flags curwin for redraw, maybe the one above, I haven't had time to look into it yet.

**Comment by bfredl** [comment]
And it is not duplicate of #3416, as K_EVENT needs not be involved: the initial case neovim/python-client#331 was an API call as part of a sync request from nvim to python, and the reproduce above doesn't involve the event loop at all: it just calls the API functions as vimL functions.

**Comment by blueyed** [comment]
I agree with @bfredl here, thanks for clearing things up!

**Comment by bfredl** [comment]
@blueyed Atually I can't reproduce. Can you reproduce from `nvim -u NONE` ?

**Comment by blueyed** [comment]
Oh, sorry.

It was apparently fixed between v0.3.0 and v0.3.1 - I must have used an older Neovim (due to previous git-bisecting likely).
